README.md
![crumb Logo](https://raw.github.com/hapijs/crumb/master/images/crumb.png)

CSRF crumb generation and validation for [**hapi**](https://github.com/hapijs/hapi)

[![Build Status](https://secure.travis-ci.org/hapijs/crumb.png)](http://travis-ci.org/hapijs/crumb)

Lead Maintainer: [Sanjay Pandit](https://github.com/spanditcaa)

## About CSRF
### What to Use Crumb for and When to Use It

Crumb is used to diminish CSRF attacks using a random unique token that is validated on the server side.

Crumb may be used whenever you want to prevent malicious code to execute system commands, that are performed by HTTP requests. For example, if users are able to publish code on your website, malicious code added by a user could force every other user who opens the page, to load and execute code from a third party website e.g. via an HTML image tag. With Crumb implemented into your hapi.js application, you are able to verify requests with unique tokens and prevent the execution of malicious requests.

### CORS

Crumb has been refactored to securely work with CORS, as [OWASP](https://www.owasp.org/index.php/HTML5_Security_Cheat_Sheet#Cross_Origin_Resource_Sharing) recommends using CSRF protection with CORS.

**It is highly discouraged to have a production servers `cors.origin` setting set to "[\*]" or "true" with Crumb as it will leak the crumb token to potentially malicious sites**


## Usage

```js
  const Hapi = require('hapi');
  const Crumb = require('crumb');

  const server = new Hapi.Server({
    port: 8000
  });

  (async () => {
    await server.register({
      plugin: Crumb,

      // plugin options
      options: {}
    });

    server.route({
      path: '/login',
      method: 'GET',
      options: {
        plugins: {
          // route specific options
          crumb: {}
        },
        handler(request, h) {
          // this requires to have a view engine configured
          return h.view('some-view');
        }
      }
    });
  })();
```

For a complete example see [the examples folder](./example).

## Options

The following options are available when registering the plugin.

### Registration options

  * `key` - the name of the cookie to store the csrf crumb into. Defaults to `crumb`.
  * `size` - the length of the crumb to generate. Defaults to `43`, which is 256 bits, see [cryptile](https://github.com/hapijs/cryptiles) for more information.
  * `autoGenerate` - whether to automatically generate a new crumb for requests. Defaults to `true`.
  * `addToViewContext` - whether to automatically add the crumb to view contexts as the given key. Defaults to `true`.
  * `cookieOptions` - storage options for the cookie containing the crumb, see the [server.state](http://hapijs.com/api#serverstatename-options) documentation of hapi for more information. Default to `cookieOptions.path=/`
  * `headerName` - specify the name of the custom CSRF header. Defaults to `X-CSRF-Token`.
  * `restful` - RESTful mode that validates crumb tokens from *"X-CSRF-Token"* request header for **POST**, **PUT**, **PATCH** and **DELETE** server routes. Disables payload/query crumb validation. Defaults to `false`.
  * `skip` - a function with the signature of `function (request, h) {}`, which when provided, is called for every request. If the provided function returns true, validation and generation of crumb is skipped. Defaults to `false`.
  * `enforce` - defaults to true, using enforce with false will set the CSRF header cookie but won't execute the validation
  * `logUnauthorized` - whether to add to the request log with tag 'crumb' and data 'validation failed' (defaults to false)
  
### Routes configuration

Additionally, some configuration can be passed on a per-route basis. Disable Crumb for a particular route by passing `false` instead of a configuration object.

  * `key` - the key used in the view contexts and payloads for the crumb. Defaults to `plugin.key`.
  * `source` - can be either `payload` or `query` specifying how the crumb will be sent in requests. Defaults to `payload`.
  * `restful` - an override for the server's 'restful' setting. Defaults to `plugin.restful`.


### Contribute

* First, install `lab` and `code` with global `npm i -g lab code`
* Run tests with `npm test`


CHANGELOG.md
Breaking changes are documented using GitHub issues, see [issues labeled "release notes"](https://github.com/hapijs/crumb/issues?q=is%3Aissue+label%3A%22release+notes%22).

If you want changes of a specific minor or patch release, you can browse the [GitHub milestones](https://github.com/hapijs/crumb/milestones?state=closed&direction=asc&sort=due_date).


example/server.js
'use strict';

const Hapi = require('hapi');
const Vision = require('vision');

const server = new Hapi.Server({
    host: '127.0.0.1',
    port: 8000
});

const plugins = [
    Vision,
    {
        plugin: require('../'),
        options: {
            cookieOptions: {
                isSecure: false
            }
        }
    }
];

(async () => {

    await server.register(plugins);

    server.views({
        relativeTo: __dirname,
        path: 'templates',
        engines: {
            html: require('handlebars')
        }
    });

    server.route({
        method: 'get',
        path: '/',
        handler: function (request, h) {

            return h.view('index', { title: 'test', message: 'hi' });
        }
    });

    server.route({
        method: 'post',
        path: '/',
        handler: function (request, h) {

            return h.view('message', { title: 'test', message: request.payload.message });
        }
    });

    await server.start();

    console.log('Example server running at:', server.info.uri);
})();


example/restful.js
'use strict';

const Hapi = require('hapi');
const Vision = require('vision');

const server = new Hapi.Server({
    host: '127.0.0.1',
    port: 8000
});

const plugins = [
    Vision,
    {
        plugin: require('../'),
        options: {
            restful: true,
            cookieOptions: {
                isSecure: false
            }
        }
    }
];

// Add Crumb plugin

(async () => {

    await server.register(plugins);

    server.route([

        // a "crumb" cookie should be set with any request
        // for cross-origin requests, set CORS "credentials" to true
        // a route returning the crumb can be created like this

        {
            method: 'GET',
            path: '/generate',
            handler: function (request, h) {

                return {
                    crumb: server.plugins.crumb.generate(request, h)
                };
            }
        },

        // request header "X-CSRF-Token" with crumb value must be set in request for this route

        {
            method: 'PUT',
            path: '/crumbed',
            handler: function (request, h) {

                return 'Crumb route';
            }
        }
    ]);

    await server.start();

    console.log('Example restful server running at:', server.info.uri);
})();


lib/index.js
'use strict';

// Load modules

const Stream = require('stream');
const Boom = require('boom');
const Cryptiles = require('cryptiles');
const Hoek = require('hoek');
const Joi = require('joi');

// Constants

const restfulValidatedMethods = ['post', 'put', 'patch', 'delete'];


// Declare internals

const internals = {};


internals.schema = Joi.object().keys({
    key: Joi.string().optional(),
    size: Joi.number().optional(),
    autoGenerate: Joi.boolean().optional(),
    addToViewContext: Joi.boolean().optional(),
    cookieOptions: Joi.object().keys(null),
    headerName: Joi.string().optional(),
    restful: Joi.boolean().optional(),
    skip: Joi.func().optional(),
    enforce: Joi.boolean().optional(),
    logUnauthorized: Joi.boolean().optional()
});


internals.defaults = {
    key: 'crumb',
    size: 43,                       // Equal to 256 bits
    autoGenerate: true,             // If false, must call request.plugins.crumb.generate() manually before usage
    addToViewContext: true,         // If response is a view, add crumb to context
    cookieOptions: {                // Cookie options (i.e. hapi server.state)
        path: '/'
    },
    headerName: 'X-CSRF-Token',     // Specify the name of the custom CSRF header
    restful: false,                 // Set to true for custom header crumb validation. Disables payload/query validation
    skip: false,                    // Set to a function which returns true when to skip crumb generation and validation,
    enforce: true,                  // Set to true for setting the CSRF cookie while not performing validation
    logUnauthorized: false          // Set to true for crumb to write an event to the request log
};

const register = (server, options) => {

    Joi.assert(options, internals.schema);

    const settings = Hoek.applyToDefaults(internals.defaults, options);

    const routeDefaults = {
        key: settings.key,
        restful: settings.restful,
        source: 'payload'
    };

    server.state(settings.key, settings.cookieOptions);

    server.ext('onPostAuth', (request, h) => {

        const unauthorizedLogger = () => {

            if (settings.logUnauthorized) {
                request.log(['crumb', 'unauthorized'], 'validation failed');
            }
        };

        const multipleCookieLogger = () => {

            request.log(['crumb'], 'multiple cookies found');
        };

        const getCrumbValue = () => {

            let crumbValue = request.plugins.crumb;

            if (Array.isArray(crumbValue)) {
                multipleCookieLogger();
                crumbValue = request.plugins.crumb[0];
            }

            return crumbValue;
        };

        // If skip function enabled. Call it and if returns true, do not attempt to do anything with crumb.

        if (settings.skip && settings.skip(request, h)) {
            return h.continue;
        }

        // Validate incoming crumb

        if (typeof request.route.settings.plugins._crumb === 'undefined') {
            if (request.route.settings.plugins.crumb ||
                !request.route.settings.plugins.hasOwnProperty('crumb') && settings.autoGenerate) {

                request.route.settings.plugins._crumb = Hoek.applyToDefaults(routeDefaults, request.route.settings.plugins.crumb || {});
            }
            else {
                request.route.settings.plugins._crumb = false;
            }
        }

        // Set crumb cookie and calculate crumb

        if ((settings.autoGenerate ||
            request.route.settings.plugins._crumb) &&
            (request.route.settings.cors ? checkCORS(request) : true)) {

            generate(request, h);
        }

        // Skip validation on dry run

        if (!settings.enforce) {
            return h.continue;
        }

        // Validate crumb

        let routeIsRestful;
        if (request.route.settings.plugins._crumb && request.route.settings.plugins._crumb.restful !== undefined) {
            routeIsRestful = request.route.settings.plugins._crumb.restful;
        }

        if (routeIsRestful === false || !routeIsRestful && settings.restful === false) {

            if (request.method !== 'post' ||
                !request.route.settings.plugins._crumb) {

                return h.continue;
            }

            const content = request[request.route.settings.plugins._crumb.source];
            if (!content || content instanceof Stream) {
                unauthorizedLogger();
                throw Boom.forbidden();
            }


            if (content[request.route.settings.plugins._crumb.key] !== getCrumbValue()) {
                unauthorizedLogger();
                throw Boom.forbidden();
            }

            // Remove crumb

            delete request[request.route.settings.plugins._crumb.source][request.route.settings.plugins._crumb.key];
        }
        else {
            if (!restfulValidatedMethods.includes(request.method) || !request.route.settings.plugins._crumb) {
                return h.continue;
            }

            const header = request.headers[settings.headerName.toLowerCase()];

            if (!header) {
                unauthorizedLogger();
                throw Boom.forbidden();
            }

            if (header !== getCrumbValue()) {
                unauthorizedLogger();
                throw Boom.forbidden();
            }

        }

        return h.continue;
    });

    server.ext('onPreResponse', (request, h) => {

        // Add to view context

        const response = request.response;

        if (settings.addToViewContext &&
            request.plugins.crumb &&
            request.route.settings.plugins._crumb &&
            !response.isBoom &&
            response.variety === 'view') {

            response.source.context = response.source.context || {};
            response.source.context[request.route.settings.plugins._crumb.key] = request.plugins.crumb;
        }

        return h.continue;
    });

    const checkCORS = function (request) {

        if (request.headers.origin) {
            return request.info.cors.isOriginMatch;
        }

        return true;
    };

    const generate = function (request, h) {

        let crumb = request.state[settings.key];
        if (!crumb) {
            crumb = Cryptiles.randomString(settings.size);
            h.state(settings.key, crumb, settings.cookieOptions);
        }

        request.plugins.crumb = crumb;
        return request.plugins.crumb;
    };

    server.expose({ generate });
};

exports.plugin = {
    pkg: require('../package.json'),
    register
};
